space-concordia-robotics/robotics-prototype
# Cmake integration 204

# Assignee Section

## Description
Uses CMake to build the Teensyduino projects. The design is inspired from unit tests, so it doesn't interact with the user, everything that is created can easily be instantly deleted. It uses out-of-source builds, to keep the folders clean. 
Lastly, quite a few things are subject to change whenever the eventual refactoring issue for the C++ code is done. An example of this is that there are quite a few non-essential projects in the workspace that are purposefully not built, so all the `continue` statements in the code are just temporary.
### Overview of design
The main CMakeLists.txt file resides in a root directory, which will be `robot/rover` in this case. It finds the Arduino/Teensy sources files on the computer, and compiles all the necessary source files into a few libraries. It recursively finds all the Arduino sketches from the root directory, and generates a `CMakeLists.txt` file for each sketch and runs it, using the `CMakeTemplate.txt.in` file, where the template take in a project name and used libraries. 
 Each sub-directory will contain a `CMakeLists.txt` file a `build` folder, with `bin` and `lib` folders in `build`, respectively containing the final `.elf` file for that specific Teensyduino project and a library created from the source codes in that sub-directory.    

### Steps for Testing

0. Make sure to have the arduino IDE + teensyduino installed (if not see steps below), as well these libraries throught the library manager. Go to `sketch -> include libraries -> manage libraries...` to open the Library manager and install the following libraries:

- - Adafruit Motor Shield V2 Library -> 1.0.9
- - ArduinoBlue -> 3.0.1
- - LSM303 by Polulu -> 3.0.1
- - SparkFun I2C GPS Reading and Control -> 1.0.2 (newest version 1.0.3 doesn't compile with teensy compiler)

1) Open `~/.bashrc`
2) Append `export ARDUINO_PATH=(<path to your Arduino IDE Folder containing the arduino executable>)`, save the changes
3) Restart terminal so changes apply, or source bashrc with `. ~/.bashrc`
4) Go to `robotics-prototype/robot/rover`
5) `mkdir build`
6) cd `build`
7) `cmake ..`
If it complains about a missing `WireIMXRT.cpp` file, it is strongly recommended to simply uninstall your arduino ide (`rm the folder`) and reinstall the newest version as well as reinstall teensyduino. Steps to do so are shown below. Anoter quick and dirty unrecommended but tested solution is to copy the contents from [github](https://github.com/PaulStoffregen/Wire/blob/master/WireIMXRT.cpp) and create the file in the Wire folder.
8) `make`

If you want to delete everything that was generated by CMake, including the directories, make sure that the `pwd` is `rover/build`

`cmake --build . --target cleanup`

The only remnant is the empty `build` folder in `robot/rover`

#### Installing arduino IDE + teensyduino from CLI

1. `export ARDUINO_PATH="$HOME/arduino_ide"`
2. `export ARDUINO_IDE_VERSION="1.8.12"`
3. `export TEENSYDUINO_VERSION="152"`
4. `wget https://downloads.arduino.cc/arduino-$ARDUINO_IDE_VERSION-linux64.tar.xz`
5. `mkdir -p $ARDUINO_PATH`
6. `tar xf arduino-$ARDUINO_IDE_VERSION-linux64.tar.xz -C $ARDUINO_PATH`
7. `curl -fsSL https://www.pjrc.com/teensy/td_$TEENSYDUINO_VERSION/TeensyduinoInstall.linux64 -o TeensyduinoInstall.linux64`
8. `chmod +x TeensyduinoInstall.linux64`
9. `./TeensyduinoInstall.linux64 --dir=$ARDUINO_PATH/arduino-$ARDUINO_IDE_VERSION`
10. `cd ~/arduino_ide/arduino-1.8.12` and `./arduino`
11. Once the IDE has opened, verify that you see options for "Teensy" under `Tools -> Board"

closes #204 

# Reviewer Section

Aside from local testing and the General Integration Test it is implied that static analysis should be included in the verification process.


For Pull Requests that do not include code changes, it is not required to perform the tests above.
